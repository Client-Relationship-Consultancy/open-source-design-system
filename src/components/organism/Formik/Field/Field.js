import React from "react"
import styled from "styled-components"
import { Field as FormikField, connect } from "formik"
import PropTypes from "prop-types"
import { colourPalette } from "../../../../brandColours"
import ErrorMessage from "../ErrorMessage"

const InputWrapper = styled.div`
  display: flex;
  flex-direction: ${props => (props.row ? "row" : "column")};
  align-items: ${props => (props.row ? "center" : "stretch")};
`
InputWrapper.displayName = "InputWrapper"

const Label = styled.label`
  font-size: 1rem;
  line-height: 100%;
`
Label.displayName = "Label"

const StyledFormikField = styled(FormikField)`
  border: none;
  border-bottom: 1px solid ${props => props.theme.black.tint40.hex};
  padding: 0.5rem 0.25rem 0.25rem;
  outline: none;
  font-size: 1rem;
  color: ${props => props.theme.black.main.hex};
  order: 2;
  margin: ${props => (props.type === "checkbox" || props.type === "radio" ? "0 0 0 0.5rem" : "0")};
  cursor: ${props => (props.type === "checkbox" || props.type === "radio" ? "pointer" : "text")};
  :active,
  :focus {
    ${props =>
      props.type !== "textarea" && `border-bottom: solid 1px ${props.theme.secondary.dark.hex}`};
    + label {
      order: 1;
      color: ${props => props.theme.secondary.dark.hex};
      font-weight: bold;
    }
  }
  ${props => props.type === "textarea" && "height: 10rem"};
  ${props => props.type === "textarea" && `border: 1px solid ${props.theme.black.tint40.hex};`};
  ${props => props.type === "textarea" && "margin-top: 0.5rem;"};
  ${props => props.type === "textarea" && "border-radius: 5px;"};
`
StyledFormikField.defaultProps = {
  theme: colourPalette.examplePalette,
}
StyledFormikField.displayName = "StyledFormikField"

const CustomField = props => {
  const { values, handleChange, handleBlur, errors, touched } = props.formik
  const { name, children, isValid, row, title, type, id, className, ...other } = props
  return (
    <div className={className}>
      <InputWrapper row={row} type={type}>
        <StyledFormikField
          id={id}
          type={type}
          component={type === "textarea" ? "textarea" : undefined}
          name={name}
          value={values && values[name] ? values[name] : ""}
          onChange={handleChange}
          onBlur={handleBlur}
          {...other}
        />
        <Label htmlFor={id}>{title}</Label>
      </InputWrapper>
      {errors && errors[name] && touched && touched[name] ? <ErrorMessage name={name} /> : null}
    </div>
  )
}
CustomField.displayName = "CustomField"

CustomField.propTypes = {
  children: PropTypes.node,
  formik: PropTypes.object,
  /** Label for the input field */
  title: PropTypes.string,
  id: PropTypes.string,
  type: PropTypes.string,
  name: PropTypes.string,
  /** Set as true to have label and input on the same row (useful for radio and checkboxes) */
  row: PropTypes.bool,
  /** Set as true to have validation available on this field */
  isValid: PropTypes.bool,
  className: PropTypes.string,
}

const Field = connect(CustomField)
Field.displayName = "Field"

export default Field
