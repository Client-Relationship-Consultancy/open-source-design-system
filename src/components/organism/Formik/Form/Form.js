import React from "react"
import styled from "styled-components"
import { Formik, Form as FormikForm } from "formik"
import PropTypes from "prop-types"

// The css selector ( > * + * ) is called a lobotomized owl.
// It selects all direct child that has a previous sibling
const StyledFormikForm = styled(FormikForm)`
  > * + * {
    margin-top: ${props => props.spacing}rem;
  }
`
StyledFormikForm.displayName = "StyledFormikForm"

// Apply displayName for static rendering in tests and storybook
FormikForm.displayName = "FormikForm"

const Form = props => {
  const {
    initialValues,
    enableReinitialize = true,
    onSubmit,
    validate,
    validationSchema,
    children,
    id,
    className,
  } = props
  return (
    <Formik
      enableReinitialize={enableReinitialize}
      initialValues={initialValues}
      onSubmit={onSubmit}
      validate={validate}
      validationSchema={validationSchema}
      id={id}
      className={className}
    >
      <StyledFormikForm spacing={props.spacing}>{children}</StyledFormikForm>
    </Formik>
  )
}
Form.displayName = "Form"

Form.defaultProps = {
  spacing: 1.5,
}

Form.propTypes = {
  /** Set initial/default values of the form */
  initialValues: PropTypes.object,
  /** Set function to handle onSubmit (Remember to set isSubmitting back to false) */
  onSubmit: PropTypes.func,
  /** Create function to validate the fields. Pass values as the paramenter, and validate with values[name]. Return an object of errors */
  validate: PropTypes.func,
  validationSchema: PropTypes.func,
  spacing: PropTypes.number,
  enableReinitialize: PropTypes.bool,
  children: PropTypes.node,
  id: PropTypes.string,
  className: PropTypes.string,
}

export default Form
