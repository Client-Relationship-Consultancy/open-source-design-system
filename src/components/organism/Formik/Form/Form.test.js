import React from "react"
import { shallow } from "enzyme"
import Form from "./Form"
import Field from "../Field"

describe("Form Component Test", () => {
  const component = shallow(
    <Form>
      <Field name="name" type="text" isValid />
      <Field name="location" type="text" row isValid />
      <Field name="subscribe" type="checkbox" row />
    </Form>,
  )
  it("Render the correct number of children Fields", () => {
    expect(component.find("StyledFormikForm").children().length).toEqual(3)
  })

  it("Match last snapshot", () => {
    expect(component).toMatchSnapshot()
  })
})
