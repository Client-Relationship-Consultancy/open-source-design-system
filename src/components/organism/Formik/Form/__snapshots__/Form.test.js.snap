// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component Test Match last snapshot 1`] = `
<Formik
  enableReinitialize={true}
  isInitialValid={false}
  validateOnBlur={true}
  validateOnChange={true}
>
  <StyledFormikForm
    spacing={1.5}
  >
    <Field
      isValid={true}
      name="name"
      type="text"
    />
    <Field
      isValid={true}
      name="location"
      row={true}
      type="text"
    />
    <Field
      name="subscribe"
      row={true}
      type="checkbox"
    />
  </StyledFormikForm>
</Formik>
`;
